client/utils/src/notification.rs

```diff
@@ -0,0 +1,172 @@
+// but WITHOUT ANY WARRANTY; without even the implied warranty of
+// MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE. See the
+// GNU General Public License for more details.
+
+// You should have received a copy of the GNU General Public License
+// along with this program. If not, see <https://www.gnu.org/licenses/>.
+
+//! Provides mpsc notification channel that can be instantiated
+//! _after_ it's been shared to the consumer and producers entities.
+//!
+//! Useful when building RPC extensions where, at service definition time, we
+//! don't know whether the specific interface where the RPC extension will be
+//! exposed is safe or not and we want to lazily build the RPC extension
+//! whenever we bind the service to an interface.
+//!
+//! See [`sc-service::builder::RpcExtensionBuilder`] for more details.
+
+use std::{marker::PhantomData, sync::Arc};
+
+use crate::mpsc::{tracing_unbounded, TracingUnboundedReceiver, TracingUnboundedSender};
+
+use parking_lot::Mutex;
+
+/// Collection of channel sending endpoints shared with the receiver side
+/// so they can register themselves.
+type SharedSenders<Payload> = Arc<Mutex<Vec<TracingUnboundedSender<Payload>>>>;
+
+/// Trait used to define the "tracing key" string used to tag
+/// and identify the mpsc channels;
+pub trait TracingKeyStr {
+	const TRACING_KEY: &'static str;
+}
+
+/// The sending half of the notifications channel(s).
+///
+/// Used to send notifications from the BEEFY gadget side.
+pub struct NotificationSender<Payload: Clone, Error> {
+	subscribers: SharedSenders<Payload>,
+	_err: PhantomData<Error>,
+}
+
+// Because of https://github.com/rust-lang/rust/issues/26925 we need to
+// manually implement `clone` here.
+impl<Payload: Clone, Error> Clone for NotificationSender<Payload, Error> {
+	fn clone(&self) -> Self {
+		Self { subscribers: self.subscribers.clone(), _err: PhantomData }
+	}
+}
+
+impl<Payload: Clone, Error> NotificationSender<Payload, Error> {
+	/// The `subscribers` should be shared with a corresponding `NotificationStream`.
+	fn new(subscribers: SharedSenders<Payload>) -> Self {
+		Self { subscribers, _err: PhantomData }
+	}
+
+	/// Send out a notification to all subscribers that a new payload is available for a
+	/// block.
+	pub fn notify(&self, payload: impl FnOnce() -> Result<Payload, Error>) -> Result<(), Error> {
+		let mut subscribers = self.subscribers.lock();
+
+		// do an initial prune on closed subscriptions
+		subscribers.retain(|n| !n.is_closed());
+
+		if !subscribers.is_empty() {
+			let payload = payload()?;
+			subscribers.retain(|n| n.unbounded_send(payload.clone()).is_ok());
+		}
+
+		Ok(())
+	}
+}
+
+/// The receiving half of the notifications channel.
+///
+/// Used to receive notifications generated at the BEEFY gadget side.
+/// The `NotificationStream` entity stores the `SharedSenders` so it can be
+/// used to add more subscriptions.
+pub struct NotificationStream<Payload: Clone, TK: TracingKeyStr, Error> {
```

Comment:
Don't we have `const` generics now? I.e. instead of passing `TrackingKeyStr` we could pass `const TracingKeyStr: &'static str` directly, no?

Reply:
FYI, right only integers, chars and bools are supported on stable for const parameters, so we can't pass an `&str` directly.

Reply:
right, const `&str` generics not supported yet, still need the associated type workaround.